doc/L2-vxlan-hld.md

```diff
@@ -0,0 +1,95 @@
+# L2 VXLAN HLD for SONiC
+
+<!-- TOC -->
+
+- [L2 VXLAN HLD for SONiC](#l2-vxlan-hld-for-sonic)
+  - [1.1. Document History](#11-document-history)
+  - [1.2. Abbreviations](#12-abbreviations)
+  - [1.3. Brief of SONIC vxlan tunnel](#13-brief-of-sonic-vxlan-tunnel)
+  - [1.4. Design considerations](#14-design-considerations)
+  - [1.5. SONiC system diagram for VXLAN](#15-sonic-system-diagram-for-vxlan)
+  - [1.6. Design changes](#16-design-changes)
+    - [1.6.1. vxlanorch changes](#161-vxlanorch-changes)
+    - [1.6.2. fdborch changes](#162-fdborch-changes)
+
+<!-- /TOC -->
+
+## 1.1. Document History
+
+Version | Date | Author | Description
+---------|----------|---------|-----
+v.01 | 03/07/2019 | Jianjun dong | Initial version
+
+## 1.2. Abbreviations
+
+Term | Definition
+---------|----------
+VXLAN | Virtual Extensible LAN
+
+## 1.3. Brief of SONIC vxlan tunnel
+
+SONIC vxlan tunnel module is contributed by Microsoft.
+
+![Figure 1.3](https://github.com/shine4chen/SONiC/blob/l2-vxlan/images/l2_vxlan_hld/L2-vxlan-HLD-figure-1.3.png)
+
+VxlanOrch is the major subsystem for Vxlan that handles configuration request. Vxlanorch creates the tunnel and attaches encap and decap mappers. Separate tunnels are created for L2 Vxlan and L3 Vxlan and can attach different VLAN/VNI or VRF/VNI to respective tunnel.
+
+Vxlan tunnel is created by static configuration currently. Vxlanmgrd VxlanOrch read the configuration from Config_DB, including VxlanTunnelTable and VxlanTunnelMapTable, then set these two tables in APP_DB and create L3 Vxlan tunnel in linux kernel. The formats of these two tables are list below.
+
+```json
+
+VXLAN_TUNNEL|{{tunnel_name}}
+    "src_ip": {{ip_address}}
+    "dst_ip": {{ip_address}} (OPTIONAL)
+
+VXLAN_TUNNEL_MAP|{{tunnel_name}}|{{tunnel_map}}
+    "vni": {{ vni_id}}
+    "vlan": {{ vlan_id }}
+
+```
+
+In layer 3 Vxlan, the tunnel map is VRF to VNI. VxlanOrch and VnetOrch, are used to support Layer 3 Vxlan completely. VNET(Virtual NETwork) is beyond this document.
+
+In layer 2 Vxlan, the tunnel map is VLAN to VNI. Layer 2 Vxlan is not supported completely, current state is TBD(To Be Done). The mainly shortcomings of L2 Vxlan are:
+
+- Not specify the encap type when creating tunnel; The encap type is MAP_TO_INVALID, must be VLAN_ID_TO_VNI.
```

Comment:
May be we should add a role attribute to config_db entry, "encap_decap", "encap_only" or "decap_only". Currently there is a use-case for L2 Vxlan  as "decap_only" tunnels and hence MAP_TO_INVALID